### PR TITLE
Adding tests for ControllerProperties and TryUpdateModel

### PR DESCRIPTION
**Pull request by harshgMSFT**

### CODE REVIEW

**Comment by harshgMSFT** [comment]
cc @rynowak

**Comment by harshgMSFT** [comment]
@rynowak ping.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs [hunk 1]
```diff
@@ -0,0 +1,601 @@
...
+namespace Microsoft.AspNet.Mvc.IntegrationTests
+{
+    public class ActionParameterIntegrationTest
+    {
+        private class Address
+        {
+            public string Street { get; set; }
+        }
+
+
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs]: extra line

File: test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs [hunk 2]
```diff
@@ -0,0 +1,601 @@
...
+            public Person3()
+            {
+                Address = new List<Address>();
+            }
+
+            public List<Address> Address { get; }
+        }
+
+        [Fact(Skip = "Extra entries in model state dictionary. #2466")]
+        public async Task ActionParameter_NonSettableCollectionModel_EmptyPrefix_GetsBound()
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs]: This one seems good

File: test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs [hunk 3]
```diff
@@ -0,0 +1,601 @@
...
+            Assert.Equal(ModelValidationState.Valid, modelState[key].ValidationState);
+        }
+
+        private class Person2
+        {
+            public List<Address> Address { get; set; }
+        }
+
+        [Fact(Skip = "Extra entries in model state dictionary. #2466")]
+        public async Task ActionParameter_SettableCollectionModel_EmptyPrefix_GetsBound()
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs]: We already have this test

File: test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs [hunk 4]
```diff
@@ -0,0 +1,601 @@
...
+    public class ActionParameterIntegrationTest
+    {
+        private class Address
+        {
+            public string Street { get; set; }
+        }
+
+
+        [Fact(Skip = "Extra entries in model state dictionary. #2466")]
+        public async Task ActionParameter_CollectionModel_EmptyPrefix_GetsBound()
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs]: We already have this test

File: test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs [hunk 5]
```diff
@@ -0,0 +1,601 @@
...
+            Assert.Equal(ModelValidationState.Valid, modelState[key].ValidationState);
+        }
+
+        private class Person6
+        {
+            public CustomReadOnlyCollection<Address> Address { get; set; }
+        }
+
+        [Fact(Skip = "Extra entries in model state dictionary. #2466")]
+        public async Task ActionParameter_ReadOnlyCollectionModel_EmptyPrefix_DoesNotGetBound()
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs]: This one seems good

File: test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs [hunk 6]
```diff
@@ -0,0 +1,601 @@
...
+            Assert.Empty(modelState.Keys);
+        }
+
+        private class Person4
+        {
+            public Address[] Address { get; set; }
+        }
+
+        [Fact(Skip = "Extra entries in model state dictionary. #2466")]
+        public async Task ActionParameter_SettableArrayModel_EmptyPrefix_GetsBound()
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs]: This one seems good

File: test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs [hunk 7]
```diff
@@ -0,0 +1,601 @@
...
+            Assert.Equal(ModelValidationState.Valid, modelState[key].ValidationState);
+        }
+
+        private class Person5
+        {
+            public Address[] Address { get; } = new Address[] { };
+        }
+
+        [Fact(Skip = "Extra entries in model state dictionary. #2466")]
+        public async Task ActionParameter_NonSettableArrayModel_EmptyPrefix_DoesNotGetBound()
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs]: :+1:

File: test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs [hunk 8]
```diff
@@ -0,0 +1,601 @@
...
+                }
+            }
+
+            System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
+            {
+                return GetEnumerator();
+            }
+        }
+
+        private Task<bool> TryUpdateModel(
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs]: used anywhere?

Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ActionParametersIntegrationTest.cs]: Nope copy paste error .. will remove.

File: test/Microsoft.AspNet.Mvc.IntegrationTests/ControllerPropertiesIntegrationTest.cs [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.Mvc.ModelBinding;
+using Xunit;
+
+namespace Microsoft.AspNet.Mvc.IntegrationTests
+{
+    public class ControllerPropertiesIntegrationTest
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ControllerPropertiesIntegrationTest.cs]: I don't see what makes these 'controller property' tests - they look just like the other ones and cover scenarios that are already covered.

Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ControllerPropertiesIntegrationTest.cs]: I agree with you. Whats your suggestion? Remove this class and add unit tests (if anything is missing?).

Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.IntegrationTests/ControllerPropertiesIntegrationTest.cs]: Yes :+1: - sorry I thought we had already discussed and that was the plan

**Comment by rynowak** [comment]
:watch: tests for `TryUpdateModel` look good.

**Comment by harshgMSFT** [comment]
@rynowak I removed the controller tests, there are already good unit tests covering the scenarios for mutable/arrayand readonly properties.

**Pull request merged**